### DISCUSSION TITLE
Inspect message object after routing

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#4589] [Q&A] [answered]
I'm forwarding messages depending on rules to other channels. Now using `var response = router.routeMessageByChannelId(channelId, message)` **just returns the response as string** but I'm interested in further investigating what in detail happened with the message in the forwarded channel.

**How could I retrieve the response object or even message object instead of a string?**
**And how could I deep-dive into the forwarded message, e.g. access it's `responseMap`?**

Many thanks for any advice.

**Comment by tonygermano** [comment]
It actually doesn't return a string. It returns a `Response` object. There is a good chance whatever method you are using to view it is showing you the `toString()` value of the object.

`router` is an instance of `VMRouter` from the User API
http://javadocs.mirthcorp.com/connect/3.11.0/user-api/com/mirth/connect/server/userutil/VMRouter.html

**Reply by tom08zehn** [reply]
True, and the response object seems to be unlinked from it's origin message. What I meant is that the only useful thing is the reponse message as string.

In other words: I just can access some properties and don't see a chance to access the origin message in order to get access to its channelMap or responseMap etc.

**Reply by tonygermano** [reply]
You can't do what you are trying to do. Messages are always passed as strings. There are no methods on the response object that can return anything but a String. The channelMap and responseMap are linked to a particular message can only be accessed during processing of that message. Once the channel sends a response message, those maps are no longer referenced anywhere and garbage collection recycles the memory.

Probably the best thing to do is serialize your data before sending it as the response as JSON or XML.

If you don't want to do this for some reason, you *can* use the globalMap for passing objects between channels, but it can get ugly. You could generate a key specific to the message to use for storing the object, then pass the key back as your response. It would then be up to the calling channel to remove that object from the globalMap so you don't have a memory leak.

The downside to doing it this way is that this data being passed through the globalMap can't be viewed in the message viewer if something goes wrong like it would have been if you would have serialized your data to a string for the response.

In the globalMap case, it might be a good idea to create an object in your globalMap that will hold all of these temporary objects instead of putting them in the globalMap directly. The Google Guava library is included with mirth, and you could use a Cache in the globalMap for holding these that would auto-expire entries after a certain amount of time. That way if something happened in the channel that prevented it from removing the key, the object would not stay in memory forever.

I would only do this if the objects were extremely large and you didn't want to keep creating copies.

**Reply by tom08zehn** [reply]
Thanks. I decided to cache some information during message processing as stringified JSON and return it as base64 encoded string (to avoid character escaping) in the MSA.3 field of the response 👍🏻

I read MSA.3 is limited to 80 chars by HL7 standards... I definitely exceed 80 chars and it fits for my purpose.

**Reply by tom08zehn** [reply]
OK the above was working for a while but cannot be applied to all kind of channels. Returning debug information in the response is not useful in all cases, e.g. if the response is sent back to the requesting system and if this is a customer/hospital system.

Different idea: you said that after processing the garbage collector cleans up the memory... How does Mirth GUI then access the mappings in the dashboard per message:

![image](https://user-images.githubusercontent.com/12202052/133447060-7a602b86-372c-49e0-9531-658c40c87282.png)

^^ Is it possible to achieve the same programatically using JS? As a starting point I'd need to know how to to identify the message after routing something to another channel because the response object doesn't contain any information such as a message hash/id.

**Reply by tonygermano** [reply]
The values are serialized before storing in the db and the message browser pulls the values from the database. Depending on the object, you may or may not get the same thing back out that you put in (if the object supports deserialization or not.) Java garbage collection will kick in for any objects that are not being referenced. 

The mirth GUI accesses that information using the Client API. I believe if you export the entire message from the GUI, the resulting XML is what the message would look like on query. This is probably not the way you want to go. If it won't work to pass info back in the response, and it won't work to store info temporarily in the globalMap, then you should probably use an external database to hold the data you are concerned with. It will likely be much easier to query.

Rather than trying to embed base64 encoded JSON in your hl7 ack, you could embed your hl7 ack in the json response (you couldn't use the auto-generated response option for this, but there is an ACK generator in the User API that can accomplish the same thing.) Then the channel that receives the response can unpack the json and send only the hl7 ack upstream if needed.

**Reply by tom08zehn** [reply]
I need to keep the HL7 channels strict and cannot request customers to extract the ACK from JSON.

Also, the debugging info is only added to the response if I switch on a certain setting. So this is more or less fine and working well for HL7 channels as I can switch off the debugging info in the MSA.3 field.

The new challenge is that other channels are returning JSON or simply nothing (RAW) and so the response cannot be an HL7 response object because it would fail (type mismatch).

If Mirth GUI is using client API then this sounds great! Because it means I could do the same programatically using JS code within Mirth. 

So my idea is to store debug information in the channelMap so it doesn't affect other messages (globalChannelMap) or even global scope (globalMap) and use the client API to get the message and then the value of channelMap variable `$c('DEBUG')`. This would work for all types of channels.

I have checked the API and `GET /channels/{channelId}/messages/{messageId}` seems to be the right thing. The missing answer is how do I know the message ID? After routing a message to a channel the message ID is not returned... at least not in the response object. I could try to trick using `GET /channels/{channelId}/messages/maxMessageId` but this is not safe - especially when running automated tests in parallel (threaded).

This brings me back to the question: **Can I replace `VMRouter.routeMessage()` by something else that returns the message ID instead of a response object?**

**Comment by tom08zehn** [comment, marked as answer]
OK, got it. So this is a true GAME CHANGER - at least for me 😅😅😅

Thanks for your help! Your comments and answers always inspire me to new ideas. I hope I can give back something to the community... so here's how I did it:

```js
/**
  1st: replace VMrouter.routeMessage() or router.routeMessageByChannelId('channelId', 'message'); by
  something that returns the message ID instead of response object, see here how to do it:
  https://github.com/nextgenhealthcare/connect/issues/4738#issuecomment-924743852
*/

// 2nd: after dispatching the message and once having the message ID, access the message's responseMap:
var messageController = com.mirth.connect.server.controllers.MessageController.getInstance();
var metaDataId = new java.lang.Integer(0); // 0 is the source connector (requires native Java Integer type)
var message_ = messageController.getMessageContent(params.channelId, messageId, [metaDataId]);
var rm = message_.getConnectorMessages().get(metaDataId).getResponseMap();
```

In the same way you can also access `.getSourceMap()` or `.getChannelMap()` 🎉 and use them to extract the data that you need.

**Reply by jonbartels** [reply]
@tom08zehn 

> I hope I can give back something to the community

You did. Someone in Slack had a related question. We were able to search and found this answer and use the information to solve their problem.
